test/extended/networking/egress_router_cni.go

```diff
@@ -0,0 +1,95 @@
+	egressRouterCNIV6Manifest = "egress-router-cni-v6-cr.yaml"
+	egressRouterCNILogs       = "/tmp/egress-router-log"
+	// Match patterns are based on what is in testdata/egress-router-cni manifests
+	ipv4MatchPattern = "IP Destinations: [80 UDP 10.100.3.0 8080 SCTP 203.0.113.26 80 8443 TCP 203.0.113.27 443]"
+	ipv6MatchPattern = "IP Destinations: [80 UDP 10:100:3::0 8080 SCTP 203:0:113::26 80 8443 TCP 203:0:113::27 443]"
+	timeOut          = 1 * time.Minute
+	interval         = 5 * time.Second
+)
+
+var _ = g.Describe("[sig-network][Feature:EgressRouterCNI]", func() {
+	oc := exutil.NewCLI(egressRouterCNIE2E)
+
+	g.It("should ensure ipv4 egressrouter cni resources are created", func() {
+		doEgressRouterCNI(egressRouterCNIV4Manifest, oc, ipv4MatchPattern)
+	})
+	InOVNKubernetesContext(
+		func() {
+			g.It("should ensure ipv6 egressrouter cni resources are created", func() {
+				doEgressRouterCNI(egressRouterCNIV6Manifest, oc, ipv6MatchPattern)
+			})
+		},
+	)
+})
+
+func doEgressRouterCNI(manifest string, oc *exutil.CLI, matchString string) error {
+	var deployment *appsv1.Deployment
+	var podList *v1.PodList
+	f := oc.KubeFramework()
+
+	g.By("creating an egressroutercni object")
+	egressRouterCNIYaml := exutil.FixturePath("testdata", "egress-router-cni", manifest)
+
+	g.By(fmt.Sprintf("calling oc create -f %s", egressRouterCNIYaml))
+	err := oc.AsAdmin().Run("create").Args("-f", egressRouterCNIYaml).Execute()
+	e2e.ExpectNoError(err, "while creating egress-router-cni object")
+	o.Expect(err).NotTo(o.HaveOccurred())
+
+	g.By("making sure egressroutercni deployment is created")
+	o.Eventually(func() error {
+		deployment, err = f.ClientSet.AppsV1().Deployments(f.Namespace.Name).Get(context.Background(), egressRouterCNIDeployment, metav1.GetOptions{})
+		return err
+	}, timeOut, interval).Should(o.Succeed())
+
+	g.By("checking network-attachment-definition resource is created")
+	output, err := oc.AsAdmin().Run("get").Args("network-attachment-definition", "-o=jsonpath={.items[0].metadata.name}").Output()
+	e2e.ExpectNoError(err, "while creating egress-router-cni network-attachement-definition")
+	o.Expect(err).NotTo(o.HaveOccurred())
+	o.Expect(output).Should(o.ContainSubstring(egressRouterCNINadObject))
+
+	g.By("getting a pod from deployment in running state")
+	o.Eventually(func() error {
+		podList, err = e2edeployment.GetPodsForDeployment(f.ClientSet, deployment)
+		return err
+	}, timeOut, interval).Should(o.Succeed())
+
+	o.Expect(podList.Items).NotTo(o.BeEmpty())
+	pod := podList.Items[0]
+	e2e.Logf("egress router cni pod %s is created\n", pod.Spec.Containers[0].Name)
+
+	node := pod.Spec.NodeName
+	expectNoError(err)
+	g.By("checking for specific output in the egress-router log on the node " + node)
+	o.Eventually(func() string {
+		result, err := oc.AsAdmin().Run("debug").Args("node/"+node, "--", "chroot", "/host", "/bin/bash", "-c", "cat /tmp/egress-router-log").Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
```

Comment:
My interpretation of the gomega documentation is that if this assert it is not done with a gomega object passed to the function in the Eventually block, and if the assert fails, Eventually will not retry and fail the test.
```
Note that the assertions in the body of the polled function must be performed using the passed-in g Gomega object. If you use the global DSL expectations, Eventually will not intercept any failures and the test will fail.
```

 Thus I would rewrite as
```
	o.Eventually(func(g o.Gomega) string {
		result, err := oc.AsAdmin().Run("debug").Args("node/"+node, "--", "chroot", "/host", "/bin/bash", "-c", "cat /tmp/egress-router-log").Output()
		g.Expect(err).NotTo(o.HaveOccurred())
		return result
	}, timeOut, interval).Should(o.ContainSubstring(matchString))
```

Reply:
this doesn't seem to work, did u find a sample from any k8s e2e doing the same ?
```
	github.com/onsi/ginkgo@v4.7.0-origin.0+incompatible/internal/leafnodes/runner.go:107 +0x96
panic({0x6fe5820, 0x8b78f90})
	runtime/panic.go:1038 +0x215
github.com/onsi/gomega/internal/asyncassertion.New(0x0, {0x711efc0, 0xc0012acf40}, 0xc000d9fe18, 0xdf8475800, 0x12a05f200, 0x0)
	github.com/onsi/gomega@v1.10.1/internal/asyncassertion/async_assertion.go:35 +0x145
github.com/onsi/gomega.EventuallyWithOffset(0xc0000ea700, {0x711efc0, 0xc0012acf40}, {0xc001fea318, 0x2, 0x1})
	github.com/onsi/gomega@v1.10.1/gomega_dsl.go:225 +0xdf
github.com/onsi/gomega.Eventually(...)
	github.com/onsi/gomega@v1.10.1/gomega_dsl.go:207
github.com/openshift/origin/test/extended/networking.doEgressRouterCNI({0x7fef78c, 0x1c}, 0xc001a4fb00, {0x8195c10, 0x58})
	github.com/openshift/origin/test/extended/networking/egress_router_cni.go:88 +0xb96
github.com/openshift/origin/test/extended/networking.glob..func2.1()
	github.com/openshift/origin/test/extended/networking/egress_router_cni.go:39 +0x35
github.com/onsi/ginkgo/internal/leafnodes.(*runner).runSync(0xc0000001a0)
	github.com/onsi/ginkgo@v4.7.0-origin.0+incompatible/internal/leafnodes/runner.go:113 +0xba
github.com/onsi/ginkgo/internal/leafnodes.(*runner).run(0xc001feaea0)
	github.com/onsi/ginkgo@v4.7.0-origin.0+incompatible/internal/leafnodes/runner.go:64 +0x125
github.com/onsi/ginkgo/internal/leafnodes.(*ItNode).Run(0x7fc45798bfff)
	github.com/onsi/ginkgo@v4.7.0-origin.0+incompatible/internal/leafnodes/it_node.go:26 +0x7b
github.com/onsi/ginkgo/internal/spec.(*Spec).runSample(0xc0033e3e00, 0xc003307268, {0x8c2a360, 0xc000422b80})
	github.com/onsi/ginkgo@v4.7.0-origin.0+incompatible/internal/spec/spec.go:215 +0x2a9
github.com/onsi/ginkgo/internal/spec.(*Spec).Run(0xc0033e3e00, {0x8c2a360, 0xc000422b80})
	github.com/onsi/ginkgo@v4.7.0-origin.0+incompatible/internal/spec/spec.go:138 +0xe7
github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpec(0xc0007ca8c0, 0xc0033e3e00)
	github.com/onsi/ginkgo@v4.7.0-origin.0+incompatible/internal/specrunner/spec_runner.go:200 +0xe5
github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpecs(0xc0007ca8c0)
	github.com/onsi/ginkgo@v4.7.0-origin.0+incompatible/internal/specrunner/spec_runner.go:170 +0x1a5
github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).Run(0xc0007ca8c0)
	github.com/onsi/ginkgo@v4.7.0-origin.0+incompatible/internal/specrunner/spec_runner.go:66 +0xc5
github.com/onsi/ginkgo/internal/suite.(*Suite).Run(0xc000323f40, {0x8c2a680, 0xc002d2c050}, {0x0, 0xc00008c000}, {0xc002c855a0, 0x1, 0x1}, {0x8d3ec18, 0xc000422b80}, ...)
	github.com/onsi/ginkgo@v4.7.0-origin.0+incompatible/internal/suite/suite.go:62 +0x4b2
github.com/openshift/origin/pkg/test/ginkgo.(*TestOptions).Run(0xc0023645a0, {0xc00145d5b0, 0xc5b09b0, 0x484dda0})
	github.com/openshift/origin/cmd/openshift-tests/openshift-tests.go:434 +0x32
github.com/openshift/origin/test/extended/util.WithCleanup(0xc0023a7c18)
	github.com/openshift/origin/test/extended/util/test.go:168 +0xad
main.newRunTestCommand.func1(0xc00202db80, {0xc00145d5b0, 0x1, 0x1})
	github.com/openshift/origin/cmd/openshift-tests/openshift-tests.go:434 +0x349
github.com/spf13/cobra.(*Command).execute(0xc00202db80, {0xc00145d560, 0x1, 0x1})
	github.com/spf13/cobra@v1.2.1/command.go:856 +0x60e
github.com/spf13/cobra.(*Command).ExecuteC(0xc00202d180)
	github.com/spf13/cobra@v1.2.1/command.go:974 +0x3bc
github.com/spf13/cobra.(*Command).Execute(...)
	github.com/spf13/cobra@v1.2.1/command.go:902
main.main.func1(0xc001b58e00)
	github.com/openshift/origin/cmd/openshift-tests/openshift-tests.go:84 +0x8a
main.main()
	github.com/openshift/origin/cmd/openshift-tests/openshift-tests.go:85 +0x3b6

fail [github.com/onsi/gomega@v1.10.1/internal/asyncassertion/async_assertion.go:35]: Test Panicked: Expected a function with no arguments and one or more return values.
```

Reply:
ok found a way to achieve the same exact thing with multiple return values
```
If multple values are returned by the function, Eventually will pass the first value to the matcher and require that all others are zero-valued. This allows you to pass Eventually a function that returns a value and an error - a common pattern in Go.
```

Reply:
> this doesn't seem to work, did u find a sample from any k8s e2e doing the same ?
> 
> ```
> 	github.com/onsi/ginkgo@v4.7.0-origin.0+incompatible/internal/leafnodes/runner.go:107 +0x96
> panic({0x6fe5820, 0x8b78f90})
> 	runtime/panic.go:1038 +0x215
> github.com/onsi/gomega/internal/asyncassertion.New(0x0, {0x711efc0, 0xc0012acf40}, 0xc000d9fe18, 0xdf8475800, 0x12a05f200, 0x0)
> 	github.com/onsi/gomega@v1.10.1/internal/asyncassertion/async_assertion.go:35 +0x145
> github.com/onsi/gomega.EventuallyWithOffset(0xc0000ea700, {0x711efc0, 0xc0012acf40}, {0xc001fea318, 0x2, 0x1})
> 	github.com/onsi/gomega@v1.10.1/gomega_dsl.go:225 +0xdf
> github.com/onsi/gomega.Eventually(...)
> 	github.com/onsi/gomega@v1.10.1/gomega_dsl.go:207
> github.com/openshift/origin/test/extended/networking.doEgressRouterCNI({0x7fef78c, 0x1c}, 0xc001a4fb00, {0x8195c10, 0x58})
> 	github.com/openshift/origin/test/extended/networking/egress_router_cni.go:88 +0xb96
> github.com/openshift/origin/test/extended/networking.glob..func2.1()
> 	github.com/openshift/origin/test/extended/networking/egress_router_cni.go:39 +0x35
> github.com/onsi/ginkgo/internal/leafnodes.(*runner).runSync(0xc0000001a0)
> 	github.com/onsi/ginkgo@v4.7.0-origin.0+incompatible/internal/leafnodes/runner.go:113 +0xba
> github.com/onsi/ginkgo/internal/leafnodes.(*runner).run(0xc001feaea0)
> 	github.com/onsi/ginkgo@v4.7.0-origin.0+incompatible/internal/leafnodes/runner.go:64 +0x125
> github.com/onsi/ginkgo/internal/leafnodes.(*ItNode).Run(0x7fc45798bfff)
> 	github.com/onsi/ginkgo@v4.7.0-origin.0+incompatible/internal/leafnodes/it_node.go:26 +0x7b
> github.com/onsi/ginkgo/internal/spec.(*Spec).runSample(0xc0033e3e00, 0xc003307268, {0x8c2a360, 0xc000422b80})
> 	github.com/onsi/ginkgo@v4.7.0-origin.0+incompatible/internal/spec/spec.go:215 +0x2a9
> github.com/onsi/ginkgo/internal/spec.(*Spec).Run(0xc0033e3e00, {0x8c2a360, 0xc000422b80})
> 	github.com/onsi/ginkgo@v4.7.0-origin.0+incompatible/internal/spec/spec.go:138 +0xe7
> github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpec(0xc0007ca8c0, 0xc0033e3e00)
> 	github.com/onsi/ginkgo@v4.7.0-origin.0+incompatible/internal/specrunner/spec_runner.go:200 +0xe5
> github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpecs(0xc0007ca8c0)
> 	github.com/onsi/ginkgo@v4.7.0-origin.0+incompatible/internal/specrunner/spec_runner.go:170 +0x1a5
> github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).Run(0xc0007ca8c0)
> 	github.com/onsi/ginkgo@v4.7.0-origin.0+incompatible/internal/specrunner/spec_runner.go:66 +0xc5
> github.com/onsi/ginkgo/internal/suite.(*Suite).Run(0xc000323f40, {0x8c2a680, 0xc002d2c050}, {0x0, 0xc00008c000}, {0xc002c855a0, 0x1, 0x1}, {0x8d3ec18, 0xc000422b80}, ...)
> 	github.com/onsi/ginkgo@v4.7.0-origin.0+incompatible/internal/suite/suite.go:62 +0x4b2
> github.com/openshift/origin/pkg/test/ginkgo.(*TestOptions).Run(0xc0023645a0, {0xc00145d5b0, 0xc5b09b0, 0x484dda0})
> 	github.com/openshift/origin/cmd/openshift-tests/openshift-tests.go:434 +0x32
> github.com/openshift/origin/test/extended/util.WithCleanup(0xc0023a7c18)
> 	github.com/openshift/origin/test/extended/util/test.go:168 +0xad
> main.newRunTestCommand.func1(0xc00202db80, {0xc00145d5b0, 0x1, 0x1})
> 	github.com/openshift/origin/cmd/openshift-tests/openshift-tests.go:434 +0x349
> github.com/spf13/cobra.(*Command).execute(0xc00202db80, {0xc00145d560, 0x1, 0x1})
> 	github.com/spf13/cobra@v1.2.1/command.go:856 +0x60e
> github.com/spf13/cobra.(*Command).ExecuteC(0xc00202d180)
> 	github.com/spf13/cobra@v1.2.1/command.go:974 +0x3bc
> github.com/spf13/cobra.(*Command).Execute(...)
> 	github.com/spf13/cobra@v1.2.1/command.go:902
> main.main.func1(0xc001b58e00)
> 	github.com/openshift/origin/cmd/openshift-tests/openshift-tests.go:84 +0x8a
> main.main()
> 	github.com/openshift/origin/cmd/openshift-tests/openshift-tests.go:85 +0x3b6
> 
> fail [github.com/onsi/gomega@v1.10.1/internal/asyncassertion/async_assertion.go:35]: Test Panicked: Expected a function with no arguments and one or more return values.
> ```
> 
> here is an example from k8s e2ehttps://github.com/kubernetes/kubernetes/blob/e31aafc4fdaa70e3e14b9402efef7bd8d153c0e5/test/e2e_node/device_plugin_test.go#L118 I couldn't find a single use case with Eventually(func( ) that use gomega as an arg in any repro :)

Ok, it seems to be related to a fix introduced in gomega 1.15
https://github.com/onsi/gomega/commit/e60e9151d39c4845da6400d00fbc48c3cafc3d98
But it is weird that we are making assertions within Eventually block because support for that was introduced in gomega 1.14 and this project is using gomega 1.10